### PR TITLE
SLCORE-31 Add support for SonarQube 4.5.x

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: core/src/main/java/org/sonarsource/sonarlint/core/container/storage/SonarQubeActiveRulesProvider.java [hunk 1]
```diff
@@ -95,4 +95,20 @@ public ActiveRules provide(Sonarlint.Rules storageRules, Sonarlint.QProfiles qPr
...
   }
 
+  private Map<String, String> loadQualityProfilesFromStorage(Sonarlint.QProfiles qProfiles, StorageManager storageManager, ConnectedAnalysisConfiguration analysisConfiguration,
+    boolean supportQualityProfilesWS) {
+    Map<String, String> qProfilesByLanguage;
+    if (analysisConfiguration.moduleKey() == null) {
+      if (!supportQualityProfilesWS) {
+        throw new UnsupportedOperationException("Unable to analyse a project with no key prior to SonarQube 5.2");
+      }
+      LOG.debug("Use default quality profiles:");
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/storage/SonarQubeActiveRulesProvider.java]: MAJOR Make "loadQualityProfilesFromStorage" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/RulesDownloader.java [hunk 2]
```diff
@@ -81,35 +95,179 @@ private void fetchRulesAndActiveRules(Rules.Builder rulesBuilder, Map<String, Ac
...
     builder.append("&ps=").append(pageSize);
     return builder.toString();
   }
 
-  private static SearchResponse loadFromStream(WsResponse response) {
-    try (InputStream is = response.contentStream()) {
-      return SearchResponse.parseFrom(is);
-    } catch (IOException e) {
-      throw new IllegalStateException("Failed to load rules", e);
+  private static SearchResponse loadFromStream(WsResponse response, Version serverVersion) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/RulesDownloader.java]: MAJOR The Cyclomatic Complexity of this method "loadFromStream" is 13 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java [hunk 3]
```diff
@@ -79,24 +86,59 @@ public void update(String moduleKey) {
...
+      while (reader.hasNext()) {
+        String propName = reader.nextName();
+        if ("qprofilesByLanguage".equals(propName)) {
+          reader.beginObject();
+          while (reader.hasNext()) {
+            String language = reader.nextName();
+            reader.beginObject();
+            while (reader.hasNext()) {
+              String qpPropName = reader.nextName();
+              switch (qpPropName) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java]: MINOR Replace this "switch" statement by "if" statements to increase readability. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1301)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java [hunk 4]
```diff
@@ -79,24 +86,59 @@ public void update(String moduleKey) {
...
+    try (JsonReader reader = new JsonReader(response.contentReader())) {
+      reader.beginObject();
+      while (reader.hasNext()) {
+        String propName = reader.nextName();
+        if ("qprofilesByLanguage".equals(propName)) {
+          reader.beginObject();
+          while (reader.hasNext()) {
+            String language = reader.nextName();
+            reader.beginObject();
+            while (reader.hasNext()) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java]: MINOR Refactor this code to not nest more than 4 if/for/while/switch/try statements. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR org.sonarsource.sonarlint.core:sonarlint-core:src/main/java/org/sonarsource/sonarlint/core/container/connected/rule: Add a 'package-info.java' file to document the 'rule' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/storage/SonarQubeActiveRulesProvider.java [hunk 5]
```diff
@@ -102,4 +102,20 @@ public ActiveRules provide(Sonarlint.Rules storageRules, Sonarlint.QProfiles qPr
     return activeRules;
   }
 
+  private Map<String, String> loadQualityProfilesFromStorage(Sonarlint.QProfiles qProfiles, StorageManager storageManager, ConnectedAnalysisConfiguration analysisConfiguration,
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/storage/SonarQubeActiveRulesProvider.java]: MAJOR Make "loadQualityProfilesFromStorage" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Comment by dbmeneses** [comment]
LGTM

**Pull request merged**